`gatt_service` Proc Macro can't handle &'static str values

Hello you all were able to help me earlier with  #272 . I'm finally looking to pick that project back up and it seems that the  `gatt_service` macro can't handle `&'static str` values. I get the following error:
```
error[E0599]: no associated item named `Value` found for struct `DeviceInfoService` in the current scope
   --> rp_firmware/src/bluetooth.rs:110:63
    |
25  | #[gatt_service(uuid = "0000180a-0000-1000-8000-00805f9b34fb")]
    | -------------------------------------------------------------- associated item `Value` not found for this struct
...
110 |             AdStructure::ServiceUuids128(&[DeviceInfoService::Value]),
    |                                                               ^^^^^ associated item not found in `DeviceInfoService`

error[E0080]: could not evaluate static initializer
  --> rp_firmware/src/bluetooth.rs:38:20
   |
38 |     serial_number: &'static str,
   |                    ^ values of the type `[u8; usize::MAX]` are too big for the target architecture

Some errors have detailed explanations: E0080, E0599.
For more information about an error, try `rustc --explain E0080`.
```

Poking around a bit it looks like the result of the `AsGatt` implmentation of `&str` setting `MAX_SIZE = usize::MAX` :
```
impl AsGatt for &'static str {
    const MIN_SIZE: usize = 0;
    const MAX_SIZE: usize = usize::MAX;

    fn as_gatt(&self) -> &[u8] {
        self.as_bytes()
    }
}
```

and the macro having the following instantiation of `StaticCell`:
```
 static #name_screaming: static_cell::StaticCell<[u8; <#ty as trouble_host::types::gatt_traits::AsGatt>::MAX_SIZE]> = static_cell::StaticCell::new();
```

It looks like this would also impact the impl for `&[u8]` since that sets `MAX_SIZE` the same way

--- a/examples/apps/src/ble_dis_peripheral.rs
+++ b/examples/apps/src/ble_dis_peripheral.rs
@@ -1,0 +1,144 @@
+use embassy_futures::join::join;
+use trouble_host::prelude::*;
+
+/// Max number of connections
+const CONNECTIONS_MAX: usize = 1;
+
+/// Max number of L2CAP channels.
+const L2CAP_CHANNELS_MAX: usize = 2; // Signal + att
+
+// GATT Server definition
+#[gatt_server]
+struct Server {
+    info_service: DeviceInformationService,
+}
+
+/// Battery service
+#[gatt_service(uuid = service::DEVICE_INFORMATION)]
+struct DeviceInformationService {
+    /// Manufacturer name
+    #[characteristic(uuid = characteristic::MANUFACTURER_NAME_STRING, read, value = "Embassy Developers")]
+    mfg_name: &'static str,
+    /// Model number
+    #[characteristic(uuid = characteristic::MODEL_NUMBER_STRING, read, value = "EX-DIS")]
+    model_number: &'static str,
+}
+
+/// Run the BLE stack.
+pub async fn run<C>(controller: C)
+where
+    C: Controller,
+{
+    // Using a fixed "random" address can be useful for testing. In real scenarios, one would
+    // use e.g. the MAC 6 byte array as the address (how to get that varies by the platform).
+    let address: Address = Address::random([0xff, 0x8f, 0x1a, 0x05, 0xe4, 0xff]);
+    info!("Our address = {:?}", address);
+
+    let mut resources: HostResources<DefaultPacketPool, CONNECTIONS_MAX, L2CAP_CHANNELS_MAX> = HostResources::new();
+    let stack = trouble_host::new(controller, &mut resources).set_random_address(address);
+    let Host {
+        mut peripheral, runner, ..
+    } = stack.build();
+
+    info!("Starting advertising and GATT service");
+    let server = Server::new_with_config(GapConfig::Peripheral(PeripheralConfig {
+        name: "TrouBLE",
+        appearance: &appearance::power_device::GENERIC_POWER_DEVICE,
+    }))
+    .unwrap();
+
+    let _ = join(ble_task(runner), async {
+        loop {
+            match advertise("Trouble Example", &mut peripheral, &server).await {
+                Ok(conn) => {
+                    // set up tasks when the connection is established to a central, so they don't run when no one is connected.
+                    let _ = gatt_events_task(&conn).await;
+                }
+                Err(e) => {
+                    #[cfg(feature = "defmt")]
+                    let e = defmt::Debug2Format(&e);
+                    panic!("[adv] error: {:?}", e);
+                }
+            }
+        }
+    })
+    .await;
+}
+
+/// This is a background task that is required to run forever alongside any other BLE tasks.
+///
+/// ## Alternative
+///
+/// If you didn't require this to be generic for your application, you could statically spawn this with i.e.
+///
+/// ```rust,ignore
+///
+/// #[embassy_executor::task]
+/// async fn ble_task(mut runner: Runner<'static, SoftdeviceController<'static>>) {
+///     runner.run().await;
+/// }
+///
+/// spawner.must_spawn(ble_task(runner));
+/// ```
+async fn ble_task<C: Controller, P: PacketPool>(mut runner: Runner<'_, C, P>) {
+    loop {
+        if let Err(e) = runner.run().await {
+            #[cfg(feature = "defmt")]
+            let e = defmt::Debug2Format(&e);
+            panic!("[ble_task] error: {:?}", e);
+        }
+    }
+}
+
+/// Stream Events until the connection closes.
+///
+/// This function will handle the GATT events and process them.
+/// This is how we interact with read and write requests.
+async fn gatt_events_task<P: PacketPool>(conn: &GattConnection<'_, '_, P>) -> Result<(), Error> {
+    let reason = loop {
+        match conn.next().await {
+            GattConnectionEvent::Disconnected { reason } => break reason,
+            GattConnectionEvent::Gatt { event } => {
+                // This step is also performed at drop(), but writing it explicitly is necessary
+                // in order to ensure reply is sent.
+                match event.accept() {
+                    Ok(reply) => reply.send().await,
+                    Err(e) => warn!("[gatt] error sending response: {:?}", e),
+                };
+            }
+            _ => {} // ignore other Gatt Connection Events
+        }
+    };
+    info!("[gatt] disconnected: {:?}", reason);
+    Ok(())
+}
+
+/// Create an advertiser to use to connect to a BLE Central, and wait for it to connect.
+async fn advertise<'values, 'server, C: Controller>(
+    name: &'values str,
+    peripheral: &mut Peripheral<'values, C, DefaultPacketPool>,
+    server: &'server Server<'values>,
+) -> Result<GattConnection<'values, 'server, DefaultPacketPool>, BleHostError<C::Error>> {
+    let mut advertiser_data = [0; 31];
+    let len = AdStructure::encode_slice(
+        &[
+            AdStructure::Flags(LE_GENERAL_DISCOVERABLE | BR_EDR_NOT_SUPPORTED),
+            AdStructure::ServiceUuids16(&[[0x0a, 0x18]]),
+            AdStructure::CompleteLocalName(name.as_bytes()),
+        ],
+        &mut advertiser_data[..],
+    )?;
+    let advertiser = peripheral
+        .advertise(
+            &Default::default(),
+            Advertisement::ConnectableScannableUndirected {
+                adv_data: &advertiser_data[..len],
+                scan_data: &[],
+            },
+        )
+        .await?;
+    info!("[adv] advertising");
+    let conn = advertiser.accept().await?.with_attribute_server(server)?;
+    info!("[adv] connection established");
+    Ok(conn)
+}

--- a/examples/apps/src/lib.rs
+++ b/examples/apps/src/lib.rs
@@ -12,6 +12,7 @@ pub mod ble_bas_central_sec;
 pub mod ble_bas_peripheral;
 pub mod ble_bas_peripheral_sec;
 pub mod ble_beacon;
+pub mod ble_dis_peripheral;
 pub mod ble_l2cap_central;
 pub mod ble_l2cap_peripheral;
 pub mod ble_scanner;

--- a/host-macros/src/characteristic.rs
+++ b/host-macros/src/characteristic.rs
@@ -48,6 +48,12 @@ pub struct AccessArgs {
     pub indicate: bool,
 }
 
+impl AccessArgs {
+    pub fn is_read_only(&self) -> bool {
+        self.read && !self.write && !self.write_without_response && !self.notify && !self.indicate
+    }
+}
+
 /// Descriptor attribute arguments.
 ///
 /// Descriptors are optional and can be used to add additional metadata to the characteristic.

--- a/host-macros/src/service.rs
+++ b/host-macros/src/service.rs
@@ -178,6 +178,31 @@ impl ServiceBuilder {
         ));
     }
 
+    /// Construct instructions for adding a read-only characteristic to the service.
+    fn construct_characteristic_ro(&mut self, characteristic: Characteristic) {
+        let (code_descriptors, named_descriptors) = self.build_descriptors(&characteristic);
+        let char_name = format_ident!("{}", characteristic.name);
+        let ty = characteristic.ty;
+        let uuid = characteristic.args.uuid;
+        let default_value = match characteristic.args.default_value {
+            Some(val) => quote!(#val),                                       // if set by user
+            None => quote_spanned!(characteristic.span => <#ty>::default()), // or default otherwise
+        };
+
+        self.code_build_chars.extend(quote_spanned! {characteristic.span=>
+            let (#char_name, #(#named_descriptors),*) = {
+                let mut builder = service.add_characteristic_ro(#uuid, #default_value);
+                #code_descriptors
+
+                (builder.build(), #(#named_descriptors),*)
+            };
+        });
+
+        self.code_struct_init.extend(quote_spanned!(characteristic.span=>
+            #char_name,
+        ));
+    }
+
     /// Consume the lists of fields and fields marked as characteristics and prepare the code to add them to the service
     /// by generating the macro blueprints for any methods, fields, and static storage required.
     pub fn process_characteristics_and_fields(
@@ -199,7 +224,19 @@ impl ServiceBuilder {
         // Process characteristic fields
         for ch in characteristics {
             let char_name = format_ident!("{}", ch.name);
-            let ty = &ch.ty;
+
+            let mut ty = &ch.ty;
+            let mut ro = false;
+            if let syn::Type::Reference(type_ref) = ty {
+                if ch.args.access.is_read_only()
+                    && type_ref.mutability.is_none()
+                    && type_ref.lifetime.as_ref().is_some_and(|lt| lt.ident == "static")
+                {
+                    ty = &type_ref.elem;
+                    ro = true;
+                }
+            }
+
             // add fields for each characteristic value handle
             fields.push(syn::Field {
                 ident: Some(char_name.clone()),
@@ -213,7 +250,11 @@ impl ServiceBuilder {
 
             self.increment_attributes(&ch.args.access);
 
-            self.construct_characteristic_static(ch);
+            if ro {
+                self.construct_characteristic_ro(ch);
+            } else {
+                self.construct_characteristic_static(ch);
+            }
         }
         assert_eq!(fields.len(), doc_strings.len());
         // Processing common to all fields

--- a/host/src/attribute.rs
+++ b/host/src/attribute.rs
@@ -10,14 +10,12 @@ use embassy_sync::blocking_mutex::Mutex;
 use heapless::Vec;
 
 use crate::att::{AttErrorCode, AttUns};
-use crate::gatt;
-
 use crate::attribute_server::AttributeServer;
 use crate::cursor::{ReadCursor, WriteCursor};
 use crate::prelude::{AsGatt, FixedGattValue, FromGatt, GattConnection};
 use crate::types::gatt_traits::FromGattError;
 pub use crate::types::uuid::Uuid;
-use crate::{Error, PacketPool, MAX_INVALID_DATA_LEN};
+use crate::{gatt, Error, PacketPool, MAX_INVALID_DATA_LEN};
 
 /// Characteristic properties
 #[derive(Debug, Clone, Copy)]
@@ -511,7 +509,7 @@ pub struct ServiceBuilder<'r, 'd, M: RawMutex, const MAX: usize> {
 }
 
 impl<'d, M: RawMutex, const MAX: usize> ServiceBuilder<'_, 'd, M, MAX> {
-    fn add_characteristic_internal<T: AsGatt>(
+    fn add_characteristic_internal<T: AsGatt + ?Sized>(
         &mut self,
         uuid: Uuid,
         props: CharacteristicProps,
@@ -591,7 +589,7 @@ impl<'d, M: RawMutex, const MAX: usize> ServiceBuilder<'_, 'd, M, MAX> {
     }
 
     /// Add a characteristic to this service with a refererence to an immutable storage buffer.
-    pub fn add_characteristic_ro<T: AsGatt, U: Into<Uuid>>(
+    pub fn add_characteristic_ro<T: AsGatt + ?Sized, U: Into<Uuid>>(
         &mut self,
         uuid: U,
         value: &'d T,
@@ -630,7 +628,7 @@ impl<M: RawMutex, const MAX: usize> Drop for ServiceBuilder<'_, '_, M, MAX> {
 /// A characteristic in the attribute table.
 #[cfg_attr(feature = "defmt", derive(defmt::Format))]
 #[derive(Clone, Copy, Debug, PartialEq)]
-pub struct Characteristic<T: AsGatt> {
+pub struct Characteristic<T: AsGatt + ?Sized> {
     /// Handle value assigned to the Client Characteristic Configuration Descriptor (if any)
     pub cccd_handle: Option<u16>,
     /// Handle value assigned to this characteristic when it is added to the Gatt Attribute Table
@@ -638,7 +636,7 @@ pub struct Characteristic<T: AsGatt> {
     pub(crate) phantom: PhantomData<T>,
 }
 
-impl<T: FromGatt> Characteristic<T> {
+impl<T: AsGatt + ?Sized> Characteristic<T> {
     /// Write a value to a characteristic, and notify a connection with the new value of the characteristic.
     ///
     /// If the provided connection has not subscribed for this characteristic, it will not be notified.
@@ -716,7 +714,10 @@ impl<T: FromGatt> Characteristic<T> {
     pub fn get<M: RawMutex, P: PacketPool, const AT: usize, const CT: usize, const CN: usize>(
         &self,
         server: &AttributeServer<'_, M, P, AT, CT, CN>,
-    ) -> Result<T, Error> {
+    ) -> Result<T, Error>
+    where
+        T: FromGatt,
+    {
         server.table().get(self)
     }
 
@@ -738,12 +739,12 @@ impl AttributeHandle for CharacteristicPropertiesHandle {
 }
 
 /// Builder for characteristics.
-pub struct CharacteristicBuilder<'r, 'd, T: AsGatt, M: RawMutex, const MAX: usize> {
+pub struct CharacteristicBuilder<'r, 'd, T: AsGatt + ?Sized, M: RawMutex, const MAX: usize> {
     handle: Characteristic<T>,
     table: &'r mut AttributeTable<'d, M, MAX>,
 }
 
-impl<'d, T: AsGatt, M: RawMutex, const MAX: usize> CharacteristicBuilder<'_, 'd, T, M, MAX> {
+impl<'d, T: AsGatt + ?Sized, M: RawMutex, const MAX: usize> CharacteristicBuilder<'_, 'd, T, M, MAX> {
     fn add_descriptor_internal<DT: AsGatt>(
         &mut self,
         uuid: Uuid,
@@ -883,7 +884,9 @@ impl CharacteristicProps {
 
 impl FixedGattValue for CharacteristicProps {
     const SIZE: usize = 1;
+}
 
+impl FromGatt for CharacteristicProps {
     fn from_gatt(data: &[u8]) -> Result<Self, FromGattError> {
         if data.len() != Self::SIZE {
             return Err(FromGattError::InvalidLength);
@@ -891,9 +894,14 @@ impl FixedGattValue for CharacteristicProps {
 
         Ok(CharacteristicProps(data[0]))
     }
+}
+
+impl AsGatt for CharacteristicProps {
+    const MIN_SIZE: usize = Self::SIZE;
+    const MAX_SIZE: usize = Self::SIZE;
 
     fn as_gatt(&self) -> &[u8] {
-        FixedGattValue::as_gatt(&self.0)
+        AsGatt::as_gatt(&self.0)
     }
 }
 

--- a/host/src/types/gatt_traits.rs
+++ b/host/src/types/gatt_traits.rs
@@ -18,18 +18,10 @@ pub enum FromGattError {
 pub trait FixedGattValue: FromGatt {
     /// Size of the type in bytes
     const SIZE: usize;
-
-    /// Converts from gatt bytes.
-    /// Must return FromGattError::InvalidLength if data.len != Self::SIZE
-    fn from_gatt(data: &[u8]) -> Result<Self, FromGattError>;
-
-    /// Converts to gatt bytes.
-    /// Must return a slice of len Self::SIZE
-    fn as_gatt(&self) -> &[u8];
 }
 
 /// Trait to allow conversion of a type to gatt bytes
-pub trait AsGatt: Sized {
+pub trait AsGatt {
     /// The minimum size the type might be
     const MIN_SIZE: usize;
     /// The maximum size the type might be
@@ -42,66 +34,69 @@ pub trait AsGatt: Sized {
 /// Trait to allow conversion of gatt bytes into a type
 ///
 /// Requires that the type implements AsGatt
-pub trait FromGatt: AsGatt {
+pub trait FromGatt: AsGatt + Sized {
     /// Converts from gatt bytes.
     /// Must return FromGattError::InvalidLength if data.len not in MIN_SIZE..=MAX_SIZE
     fn from_gatt(data: &[u8]) -> Result<Self, FromGattError>;
 }
 
-impl<T: FixedGattValue> FromGatt for T {
-    fn from_gatt(data: &[u8]) -> Result<Self, FromGattError> {
-        <Self as FixedGattValue>::from_gatt(data)
-    }
+macro_rules! primitive {
+    ($($ty:ty),*) => {
+        $(
+            impl FixedGattValue for $ty {
+                const SIZE: usize = mem::size_of::<Self>();
+            }
+
+            impl FromGatt for $ty {
+                fn from_gatt(data: &[u8]) -> Result<Self, FromGattError> {
+                    if data.len() != Self::SIZE {
+                        Err(FromGattError::InvalidLength)
+                    } else {
+                        // SAFETY
+                        // - Pointer is considered "valid" as per the rules outlined for validity in std::ptr v1.82.0
+                        // - Pointer was generated from a slice of bytes matching the size of the type, and all primitives are valid for all possible configurations of bits
+                        // - Primitives are copy
+                        unsafe { Ok((data.as_ptr() as *const Self).read_unaligned()) }
+                    }
+                }
+            }
+
+            impl AsGatt for $ty {
+                const MIN_SIZE: usize = mem::size_of::<Self>();
+                const MAX_SIZE: usize = mem::size_of::<Self>();
+
+                fn as_gatt(&self) -> &[u8] {
+                    // SAFETY
+                    // - Slice is of type u8 so data is guaranteed valid for reads of any length
+                    // - Data and len are tied to the address and size of the type
+                    unsafe { slice::from_raw_parts(self as *const Self as *const u8, Self::SIZE) }
+                }
+            }
+        )*
+    };
 }
 
-impl<T: FixedGattValue> AsGatt for T {
-    const MIN_SIZE: usize = Self::SIZE;
-    const MAX_SIZE: usize = Self::SIZE;
-
-    fn as_gatt(&self) -> &[u8] {
-        <Self as FixedGattValue>::as_gatt(self)
-    }
-}
-
-trait Primitive: Copy {}
-impl Primitive for u8 {}
-impl Primitive for u16 {}
-impl Primitive for u32 {}
-impl Primitive for u64 {}
-impl Primitive for i8 {}
-impl Primitive for i16 {}
-impl Primitive for i32 {}
-impl Primitive for i64 {}
-impl Primitive for f32 {}
-impl Primitive for f64 {}
-impl Primitive for BluetoothUuid16 {} // ok as this is just a NewType(u16)
-
-impl<T: Primitive> FixedGattValue for T {
-    const SIZE: usize = mem::size_of::<Self>();
-
-    fn from_gatt(data: &[u8]) -> Result<Self, FromGattError> {
-        if data.len() != Self::SIZE {
-            Err(FromGattError::InvalidLength)
-        } else {
-            // SAFETY
-            // - Pointer is considered "valid" as per the rules outlined for validity in std::ptr v1.82.0
-            // - Pointer was generated from a slice of bytes matching the size of the type implementing Primitive, and all types implementing Primitive are valid for all possible configurations of bits
-            // - Primitive trait is constrained to require Copy
-            unsafe { Ok((data.as_ptr() as *const Self).read_unaligned()) }
-        }
-    }
-
-    fn as_gatt(&self) -> &[u8] {
-        // SAFETY
-        // - Slice is of type u8 so data is guaranteed valid for reads of any length
-        // - Data and len are tied to the address and size of the type
-        unsafe { slice::from_raw_parts(self as *const Self as *const u8, Self::SIZE) }
-    }
-}
+primitive!(
+    u8,
+    u16,
+    u32,
+    u64,
+    u128,
+    i8,
+    i16,
+    i32,
+    i64,
+    i128,
+    f32,
+    f64,
+    BluetoothUuid16
+);
 
 impl FixedGattValue for bool {
     const SIZE: usize = 1;
+}
 
+impl FromGatt for bool {
     fn from_gatt(data: &[u8]) -> Result<Self, FromGattError> {
         if data.len() != Self::SIZE {
             Err(FromGattError::InvalidLength)
@@ -109,6 +104,11 @@ impl FixedGattValue for bool {
             Ok(data != [0x00])
         }
     }
+}
+
+impl AsGatt for bool {
+    const MIN_SIZE: usize = Self::SIZE;
+    const MAX_SIZE: usize = Self::SIZE;
 
     fn as_gatt(&self) -> &[u8] {
         match self {
@@ -170,8 +170,9 @@ impl<const N: usize> AsGatt for String<N> {
     }
 }
 
-impl AsGatt for &'static str {
+impl AsGatt for str {
     const MIN_SIZE: usize = 0;
+
     const MAX_SIZE: usize = usize::MAX;
 
     fn as_gatt(&self) -> &[u8] {
@@ -179,8 +180,9 @@ impl AsGatt for &'static str {
     }
 }
 
-impl AsGatt for &'static [u8] {
+impl AsGatt for [u8] {
     const MIN_SIZE: usize = 0;
+
     const MAX_SIZE: usize = usize::MAX;
 
     fn as_gatt(&self) -> &[u8] {
@@ -200,5 +202,15 @@ impl AsGatt for crate::types::uuid::Uuid {
 impl FromGatt for crate::types::uuid::Uuid {
     fn from_gatt(data: &[u8]) -> Result<Self, FromGattError> {
         Self::try_from(data).map_err(|_| FromGattError::InvalidLength)
+    }
+}
+
+impl<T: AsGatt + ?Sized> AsGatt for &T {
+    const MIN_SIZE: usize = T::MIN_SIZE;
+
+    const MAX_SIZE: usize = T::MAX_SIZE;
+
+    fn as_gatt(&self) -> &[u8] {
+        <T as AsGatt>::as_gatt(*self)
     }
 }

